### PR TITLE
Expose instance lifecycle events on ReactNativeHost

### PR DESCRIPTION
**Pull request by acoates-ms**
This exposes 3 events on `ReactNativeHost` for instance lifecycle events.

These events are:
 - **InstanceCreated**: Once the JS instance is created, but before the JS bundle has been loaded into it.
 - **InstanceLoaded**: After the JS bundle has been loaded. - May return that the instance failed to load
 - **InstanceDestroyed**: When the instance is destroyed.

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/react-native-windows/pull/5631)

### CODE REVIEW

File: vnext/Microsoft.ReactNative/ReactNativeHost.idl [hunk 1]
```diff
@@ -20,5 +36,9 @@ namespace Microsoft.ReactNative {
     Windows.Foundation.IAsyncAction LoadInstance();
     Windows.Foundation.IAsyncAction ReloadInstance();
     Windows.Foundation.IAsyncAction UnloadInstance();
+
+    event Windows.Foundation.EventHandler<InstanceCreatedEventArgs> InstanceCreated;
```
Comment (vmoroz) [inline vnext/Microsoft.ReactNative/ReactNativeHost.idl:40]: >InstanceCreated [](start = 68, length = 15)

I was thinking that we will expose them on the ReactInstanceSettings class instead.

File: vnext/Microsoft.ReactNative/ReactNativeHost.cpp [hunk 2]
```diff
@@ -24,6 +25,31 @@ using namespace xaml::Controls;
 
 namespace winrt::Microsoft::ReactNative::implementation {
 
+InstanceCreatedEventArgs::InstanceCreatedEventArgs(Mso::CntPtr<Mso::React::IReactContext> &&context)
+    : m_context(nullptr /*winrt::make<winrt::Microsoft::ReactNative::IReactContext>(std::move(context))*/) {}
```
Comment (vmoroz) [inline vnext/Microsoft.ReactNative/ReactNativeHost.cpp:29]: >nullptr [](start = 16, length = 7)

why is it nullptr?

File: vnext/Microsoft.ReactNative/ReactHost/React.h [hunk 3]
```diff
@@ -45,9 +46,10 @@ enum class LogLevel : int32_t {
 
 using OnErrorCallback = Mso::Functor<void(const Mso::ErrorCode &)>;
 using OnLoggingCallback = Mso::Functor<void(LogLevel logLevel, const char *message)>;
-using OnReactInstanceCreatedCallback = Mso::Functor<void(IReactInstance &)>;
-using OnReactInstanceLoadedCallback = Mso::Functor<void(IReactInstance &, const Mso::ErrorCode &)>;
-using OnReactInstanceDestroyedCallback = Mso::Functor<void(IReactInstance &)>;
+using OnReactInstanceCreatedCallback = Mso::Functor<void(Mso::CntPtr<IReactContext> &&, IReactInstance &)>;
```
Comment (vmoroz) [inline vnext/Microsoft.ReactNative/ReactHost/React.h:49]: >, IReactInstance & [](start = 86, length = 18)

Should we completely remove it in favor of IReactContext?

Comment (acoates-ms) [reply vnext/Microsoft.ReactNative/ReactHost/React.h:49]: At some point.  I think its being used in Office right now, so I wanted to keep the old interface until I was sure we didn't need it there.

File: vnext/Microsoft.ReactNative/ReactNativeHost.cpp (hunk 2 above)
Comment (acoates-ms) [reply vnext/Microsoft.ReactNative/ReactNativeHost.cpp:29]: urgh.. work in progress

File: vnext/Microsoft.ReactNative/ReactNativeHost.idl (hunk 1 above)
Comment (acoates-ms) [reply vnext/Microsoft.ReactNative/ReactNativeHost.idl:40]: The difference being that they would be snapshotted on instance creation, right?

That kind of makes sense, but then the user can no longer unregister very easily. - Thoughts?

**Pull request merged**